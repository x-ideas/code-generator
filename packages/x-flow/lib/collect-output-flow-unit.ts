/*
 * @name:
 * description:
 *  用于收集所有子flow unit的output的数据，统一传递给下一个flow unit
 */

import { XGroupedFlowUnit } from './grouped-flow-unit';

export class CollectOutputFlowUnit extends XGroupedFlowUnit {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
  async doWork(data: any): Promise<any[]> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result: any[] = [];

    let nextData = data;
    for (const unit of this._units) {
      nextData = await unit.doWork(nextData);
      result.push(nextData);
    }

    return result;
  }
}
