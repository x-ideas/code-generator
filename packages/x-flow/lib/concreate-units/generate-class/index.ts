/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  将interface生成前端class定义的FlowUnit
 */

import * as babelParser from '@babel/parser';
import lodash from 'lodash';

import { XFlowUnit } from '../../flow-unit';

import jscodeshift from 'jscodeshift';
import { Collection } from 'jscodeshift/src/Collection';
import * as K from 'jscodeshift/node_modules/ast-types/gen/kinds';
import { parserConfig } from '../../utils/jscodeshift-parser';

export class GenerateClassFlowUnit extends XFlowUnit {
  #convertedFromNiceFormat: boolean;
  constructor({ convertedFromNiceFormat = true }: { convertedFromNiceFormat: boolean }) {
    super();

    this.#convertedFromNiceFormat = convertedFromNiceFormat;
  }

  /**
   *
   * @private
   * @param {Collection<jscodeshift.>} collection
   * @returns {Collection<ASTNode>}
   * @memberof GenerateClassFlowUnit
   */
  private _changeInterfaceToClass(collection: Collection<jscodeshift.ASTNode>): Collection<jscodeshift.ASTNode> {
    const newCollection = jscodeshift('');
    const newProgram = newCollection.find(jscodeshift.Program);

    collection.find(jscodeshift.TSInterfaceDeclaration).forEach(path => {
      const classDeclaration = generateClassDeclaration(path, {
        convertedFromNiceFormat: this.#convertedFromNiceFormat,
      });
      newProgram.get('body').value.push(classDeclaration);
    });

    return newCollection;
  }

  async doWork(interfaceSources: string): Promise<string> {
    const collection: Collection<jscodeshift.ASTNode> = jscodeshift(interfaceSources, {
      parser: parserConfig(),
    });
    // const a = collection.find(jscodeshift.TSInterfaceDeclaration).get('body').get('body');
    // // console.log(collection.find(jscodeshift.TSInterfaceDeclaration).length);

    // // const ds = collection.find(jscodeshift.TSInterfaceDeclaration);
    // // const id = ds.get('id');
    // console.log(collection.find(jscodeshift.TSInterfaceDeclaration).get('id').get('name').value);

    return this._changeInterfaceToClass(collection).toSource();
  }
}

/**
 * 生成exporse装饰器
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {jscodeshift.Decorator}
 */
function generateExposeDecorator(
  property: jscodeshift.TSPropertySignature,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.Decorator {
  // 找到参数
  const argums: jscodeshift.ObjectExpression[] = [];
  if (jscodeshift.Identifier.check(property.key)) {
    // property

    const bName: string = options.convertedFromNiceFormat ? lodash.snakeCase(property.key.name) : property.key.name;

    const nameProperty = jscodeshift.property('init', jscodeshift.identifier('name'), jscodeshift.literal(bName));
    argums.push(jscodeshift.objectExpression([nameProperty]));
  }

  return jscodeshift.decorator(jscodeshift.callExpression(jscodeshift.identifier('Expose'), argums));
}

/**
 * 生成validate装饰器
 * TODO: 判断是否是可选的属性
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {jscodeshift.Decorator}
 */
function generateValidateDecorator(property: jscodeshift.TSPropertySignature): jscodeshift.Decorator | null {
  // let name: string = '';
  // if (jscodeshift.Identifier.check(property.key)) {
  //   name = property.key.name;
  // }

  // if (!name) {
  //   return null;
  // }

  if (property.optional) {
    return null;
  }

  if (property.typeAnnotation && jscodeshift.TSTypeAnnotation.check(property.typeAnnotation)) {
    // TODO: 根据类型和name生成对应的call expression

    switch (property.typeAnnotation.typeAnnotation.type) {
      // number, enum当作number
      case 'TSNumberKeyword': {
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(
              jscodeshift.identifier.from({
                comments: [jscodeshift.commentBlock('NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型')],
                name: 'getDefaultValidNum',
              }),
              []
            ),
          ])
        );

        break;
      }

      case 'TSStringKeyword': {
        // 判断是否为id // string;
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(jscodeshift.identifier('getDefaultValidStr'), []),
          ])
        );
      }

      case 'TSBooleanKeyword':
        // boolean
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(jscodeshift.identifier('getDefaultValidBoolean'), []),
          ])
        );

      case 'TSTypeReference':
        // 另一个类型定义(interface, class, enum),
        // NOTE: 通过前缀区分
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('Type'), [
            // @ts-ignore
            jscodeshift.arrowFunctionExpression([], jscodeshift.identifier(property.typeAnnotation.typeAnnotation.typeName?.name)),
          ])
        );

      case 'TSArrayType':
        {
          if (property.typeAnnotation.typeAnnotation.elementType.type === 'TSTypeReference') {
            const typeName = property.typeAnnotation.typeAnnotation.elementType.typeName;
            if (jscodeshift.Identifier.check(typeName)) {
              return jscodeshift.decorator(
                jscodeshift.callExpression(jscodeshift.identifier('Type'), [
                  jscodeshift.arrowFunctionExpression([], jscodeshift.identifier(typeName.name)),
                ])
              );
            }
          }
        }

        break;

      default:
        break;
    }
  }

  return null;
}

/**
 * 生成class property的初始化值
 * @param {jscodeshift.TSPropertySignature} property
 */
function generateClassPropertyInitValue(property: jscodeshift.TSPropertySignature): K.ExpressionKind | null {
  // 找到type类型
  const optional = property.optional;
  if (optional) {
    return null;
  }

  switch (property.typeAnnotation?.typeAnnotation.type) {
    case 'TSBooleanKeyword':
      return jscodeshift.literal(false);

    case 'TSNumberKeyword':
      return jscodeshift.literal(0);

    case 'TSStringKeyword': {
      // 是否是id
      jscodeshift.Identifier.assert(property.key);
      if (jscodeshift.Identifier.check(property.key)) {
        if (property.key.name.includes('id')) {
          return jscodeshift.callExpression(jscodeshift.identifier('getFrontId'), []);
        }
      }

      return jscodeshift.stringLiteral('');
    }

    case 'TSArrayType':
      return jscodeshift.arrayExpression([]);

    case 'TSTypeReference': {
      // TODO: 根据名字去判断
      return null;
    }

    default:
      return null;
  }
}

/**
 * 生成class property的标注
 * 如果property是可选的，那么会得到一个联合类型
 * @param {jscodeshift.TSParameterProperty} property
 * @returns {TSTypeAnnotationKind}
 */
function generateClassPropertyAnnotation(property: jscodeshift.TSPropertySignature): jscodeshift.TSTypeAnnotation | undefined {
  function _generate(typeAnnotation: K.TSTypeKind | K.TSTypeAnnotationKind, optional = false): jscodeshift.TSTypeAnnotation | undefined {
    switch (typeAnnotation.type) {
      case 'TSBooleanKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsBooleanKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsBooleanKeyword());
      }
      case 'TSNumberKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsNumberKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsNumberKeyword());
      }
      case 'TSStringKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsStringKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsStringKeyword());
      }

      case 'TSArrayType': {
        return optional
          ? jscodeshift.tsTypeAnnotation(
              jscodeshift.tsUnionType([jscodeshift.tsArrayType(typeAnnotation.elementType), jscodeshift.tsUndefinedKeyword()])
            )
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsArrayType(typeAnnotation.elementType));
      }

      case 'TSTypeReference': {
        return optional
          ? jscodeshift.tsTypeAnnotation(
              jscodeshift.tsUnionType([jscodeshift.tsTypeReference(typeAnnotation.typeName), jscodeshift.tsUndefinedKeyword()])
            )
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(typeAnnotation.typeName));
      }
    }
    return undefined;
  }

  if (property.typeAnnotation) {
    switch (property.typeAnnotation.type) {
      case 'TSTypeAnnotation': {
        return _generate(property.typeAnnotation.typeAnnotation, property.optional);
        break;
      }

      default:
        break;
    }
  }

  return undefined;
}

/**
 * 生成class的属性定义
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {(jscodeshift.ClassProperty | undefined)}
 */
function generateClassPropertyFromInterfaceProperty(
  property: jscodeshift.TSPropertySignature,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.ClassProperty | null {
  if (jscodeshift.Identifier.check(property.key)) {
    const ccName = lodash.camelCase(property.key.name);

    const value = generateClassPropertyInitValue(property);
    const typeAnnotation = generateClassPropertyAnnotation(property);
    const cProperty = jscodeshift.classProperty(jscodeshift.identifier(ccName), value, typeAnnotation);
    cProperty.comments = property.comments;

    // @ts-ignore
    // NOTE: 定义居然没有
    cProperty.decorators = [generateExposeDecorator(property, options), generateValidateDecorator(property)].filter(Boolean);

    return cProperty;
  }
  return null;
}

/**
 * 生成class声明
 * @param {string} className
 * @param {jscodeshift.TSPropertySignature[]} properties
 * @returns {jscodeshift.ExportNamedDeclaration}
 */
function generateClassDeclaration(
  interfaceDeclaration: jscodeshift.ASTPath<jscodeshift.TSInterfaceDeclaration>,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.ExportNamedDeclaration {
  const interfaceName: string = jscodeshift(interfaceDeclaration).get('id', 'name').value;
  const properties = jscodeshift(interfaceDeclaration).get('body', 'body').value;

  // @ts-ignore
  const cProperties = properties.map(prop => generateClassPropertyFromInterfaceProperty(prop, options));

  // 将第一个I替换成C
  return jscodeshift.exportNamedDeclaration(
    jscodeshift.classDeclaration(jscodeshift.identifier(`C${interfaceName.slice(1)}`), jscodeshift.classBody(cProperties))
  );
}
