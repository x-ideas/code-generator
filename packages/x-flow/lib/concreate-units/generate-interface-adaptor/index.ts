import { XFlowUnit } from '../../flow-unit';

import * as babelParser from '@babel/parser';
import lodash from 'lodash';

import jscodeshift, { ASTPath } from 'jscodeshift';
import { Collection } from 'jscodeshift/src/Collection';

interface IGenerateInterfaceAdaptorOptions {
  fromType: string;
  from: string;

  toType: string;
  // to: string;

  isConvertedFromFront: boolean;
}

export class GenerateInterfaceAdaptor extends XFlowUnit {
  _getFuncName(name: string, options: IGenerateInterfaceAdaptorOptions) {
    return `to${options.isConvertedFromFront ? 'B' : 'F'}${name}Adaptor`;
  }

  _generateInterfaceAdaptor(
    objectName: string,
    options: IGenerateInterfaceAdaptorOptions,
    properties: jscodeshift.TSPropertySignature[]
  ): jscodeshift.Property[] {
    const result: jscodeshift.Property[] = [];

    for (const property of properties) {
      if (jscodeshift.Identifier.check(property.key)) {
        const name = property.key.name;

        let isRef = false;
        let typeName = '';
        let isArray = false;

        if (jscodeshift.TSTypeReference.check(property.typeAnnotation?.typeAnnotation)) {
          isRef = true;
          if (jscodeshift.Identifier.check(property.typeAnnotation?.typeAnnotation.typeName))
            typeName = property.typeAnnotation?.typeAnnotation.typeName.name ?? '';
        } else if (jscodeshift.TSArrayType.check(property.typeAnnotation?.typeAnnotation)) {
          isArray = true;

          if (jscodeshift.TSTypeReference.check(property.typeAnnotation?.typeAnnotation.elementType)) {
            isRef = true;
            if (jscodeshift.Identifier.check(property.typeAnnotation?.typeAnnotation.elementType.typeName))
              typeName = property.typeAnnotation?.typeAnnotation.elementType.typeName.name ?? '';
          }
        }

        const memberExpression = jscodeshift.memberExpression(jscodeshift.identifier(objectName), jscodeshift.identifier(name));
        result.push(
          jscodeshift.property(
            'init',
            jscodeshift.identifier(options.isConvertedFromFront ? lodash.snakeCase(name) : lodash.camelCase(name)),
            isArray
              ? isRef
                ? jscodeshift.callExpression(memberExpression, [
                    jscodeshift.arrowFunctionExpression(
                      [jscodeshift.identifier('item')],
                      jscodeshift.callExpression(jscodeshift.identifier(this._getFuncName(typeName, options)), [jscodeshift.identifier('item')])
                    ),
                  ])
                : memberExpression
              : isRef
              ? jscodeshift.callExpression(jscodeshift.identifier(this._getFuncName(typeName, options)), [memberExpression])
              : memberExpression
          )
        );
      }
    }

    return result;
  }

  async doWork(options: IGenerateInterfaceAdaptorOptions): Promise<string[]> {
    const collection: Collection<jscodeshift.ASTNode> = jscodeshift(options.from, {
      parser: {
        parse(source: string) {
          return babelParser.parse(source, {
            sourceType: 'module',
            // 支持typescript, jsx
            plugins: [
              'estree',
              'typescript',
              [
                'decorators',
                {
                  decoratorsBeforeExport: true,
                },
              ],
              'exportDefaultFrom',
              'classProperties',
              'classPrivateProperties',
            ],
          });
        },
      },
    });

    const funcStr: string[] = [];
    collection.find(jscodeshift.TSInterfaceDeclaration).forEach((path: ASTPath<jscodeshift.TSInterfaceDeclaration>) => {
      if (jscodeshift.Identifier.check(path.node.id)) {
        const interfaceName = path.node.id.name;

        const interfaceProperties: jscodeshift.TSPropertySignature[] = path.node.body.body as jscodeshift.TSPropertySignature[];

        const func = jscodeshift.functionDeclaration(
          jscodeshift.identifier(this._getFuncName(interfaceName, options)),
          [
            jscodeshift.identifier.from({
              name: options.isConvertedFromFront ? 'fInfo' : 'bInfo',
              typeAnnotation: jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(jscodeshift.identifier(options.fromType))),
            }),
          ],
          jscodeshift.blockStatement([])
        );
        func.returnType = jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(jscodeshift.identifier(options.toType)));

        const result: Collection<any> = jscodeshift(func);

        const functionBody = func.body.body;

        functionBody.push(
          jscodeshift.returnStatement(
            jscodeshift.objectExpression(
              this._generateInterfaceAdaptor(options.isConvertedFromFront ? 'fInfo' : 'bInfo', options, interfaceProperties)
            )
          )
        );

        funcStr.push(result.toSource());
      }
    });

    return funcStr;
  }
}
