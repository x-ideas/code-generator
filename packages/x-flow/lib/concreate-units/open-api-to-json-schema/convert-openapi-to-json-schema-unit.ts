/*
 * @name:
 * description:
 *  将openAPI转换成JSON Schema的处理过程
 */

import { JSONSchema4 } from 'json-schema';
import { OpenAPIV2 } from 'openapi-types';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import toJsonSchema from '@openapi-contrib/openapi-schema-to-json-schema';

import { XFlowUnit } from '../../flow-unit';

export class ConvertOpenAPIToJsonSchemeFlowUnit extends XFlowUnit {
  async doWork(openApi: OpenAPIV2.Document): Promise<JSONSchema4> {
    return toJsonSchema(openApi);
  }
}
