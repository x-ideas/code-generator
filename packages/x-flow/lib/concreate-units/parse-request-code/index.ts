export * from './parser-request-code-unit';
