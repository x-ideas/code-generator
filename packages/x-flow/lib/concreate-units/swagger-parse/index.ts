export * from './swagger-parser-flow-unit';
