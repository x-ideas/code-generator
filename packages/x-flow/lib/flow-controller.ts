import { XFlowUnit } from './flow-unit';

type FuncType = (...args: unknown[]) => unknown;

function compose(funcs: FuncType[]): FuncType {
  return funcs.reduce((accum, current) => {
    return async function (data: unknown) {
      const nextData: any = await accum(data);
      return current(nextData);
    };
  });
}

export class XFlowController {
  private _units: XFlowUnit[];

  constructor() {
    this._units = [];
  }

  addUnit(unit: XFlowUnit): void {
    this._units.push(unit);
  }

  async run(data?: any): Promise<any> {
    const final = compose(this._units.map(unit => unit.doWork.bind(unit)));

    const result = final(data);
    return result;
  }
}
