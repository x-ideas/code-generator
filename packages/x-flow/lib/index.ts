export * from './concreate-units/generate-service-request';
