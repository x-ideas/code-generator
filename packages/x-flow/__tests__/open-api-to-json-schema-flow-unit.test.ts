import { OpenAPIV2 } from 'openapi-types';
import { SwaggerParseFlowUnit } from '../lib/concreate-units/swagger-parse';
import { ParseRequestCodeFlowUnit } from '../lib/concreate-units/parse-request-code';
import OpenAPIData from './assets/openAPI.json';
import { XFlowController } from '../lib/flow-controller';
import { CollectOutputFlowUnit } from '../lib/collect-output-flow-unit';
import { SpecialOutputFlowUnit } from '../lib/concreate-units/specify-output';
import { ConvertOpenAPIToJsonSchemeFlowUnit } from '../lib/concreate-units/open-api-to-json-schema/convert-openapi-to-json-schema-unit';
import { JSONSchema4 } from 'json-schema';

describe('测试 open api to json schema', () => {
  it('解析3610401', async () => {
    const fc = new XFlowController();

    const collectOutputUnit = new CollectOutputFlowUnit();

    const outputCodeUnit = new SpecialOutputFlowUnit('3610401');
    const outputOpenAPIUnit = new SpecialOutputFlowUnit(OpenAPIData);
    collectOutputUnit.addUnit(outputCodeUnit);
    collectOutputUnit.addUnit(outputOpenAPIUnit);

    const codeParseUnit = new ParseRequestCodeFlowUnit();
    const swaggerParseUnit = new SwaggerParseFlowUnit();

    const openAPIToJsonSchemaUnit = new ConvertOpenAPIToJsonSchemeFlowUnit();

    fc.addUnit(collectOutputUnit);
    fc.addUnit(codeParseUnit);
    fc.addUnit(swaggerParseUnit);
    fc.addUnit(openAPIToJsonSchemaUnit);

    const result: JSONSchema4 = await fc.run();
    expect(result).toMatchSnapshot();
  });
});
