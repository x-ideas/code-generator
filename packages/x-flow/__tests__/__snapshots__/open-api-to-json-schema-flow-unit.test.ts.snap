// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试 open api to json schema 解析3610401 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "definitions": Object {
    "Pager": Object {
      "properties": Object {
        "first_item_on_page": Object {
          "format": "int32",
          "type": "integer",
        },
        "has_next_page": Object {
          "type": "boolean",
        },
        "has_previous_page": Object {
          "type": "boolean",
        },
        "is_first_page": Object {
          "type": "boolean",
        },
        "is_last_page": Object {
          "type": "boolean",
        },
        "last_item_on_page": Object {
          "format": "int32",
          "type": "integer",
        },
        "page_count": Object {
          "format": "int32",
          "type": "integer",
        },
        "page_number": Object {
          "format": "int32",
          "type": "integer",
        },
        "page_size": Object {
          "format": "int32",
          "type": "integer",
        },
        "total_item_count": Object {
          "format": "int64",
          "type": "integer",
        },
      },
      "title": "Pager",
      "type": "object",
    },
    "PupuPagerDataResponse«List«补货需求单，列表用»»": Object {
      "properties": Object {
        "count": Object {
          "format": "int64",
          "type": "integer",
        },
        "data": Object {
          "items": Object {
            "items": Object {
              "properties": Object {
                "apply_no": Object {
                  "description": "需求单号",
                  "type": "string",
                },
                "apply_type": Object {
                  "description": "补货类型,10=旧版采购补货,20=总仓自动补货,30=DC仓采购补货(手动)",
                  "format": "int32",
                  "type": "integer",
                },
                "date_part_plan": Object {
                  "description": "计划到货时间段，10=白天，20=晚上",
                  "format": "int32",
                  "type": "integer",
                },
                "date_plan": Object {
                  "description": "计划到货日期",
                  "format": "int64",
                  "type": "integer",
                },
                "dc_abbr_name": Object {
                  "description": "总仓简称",
                  "type": "string",
                },
                "dc_id": Object {
                  "description": "总仓id",
                  "type": "string",
                },
                "dc_name": Object {
                  "description": "总仓名称",
                  "type": "string",
                },
                "id": Object {
                  "description": "id",
                  "type": "string",
                },
                "is_delete": Object {
                  "description": "是否删除",
                  "example": false,
                  "type": "boolean",
                },
                "item_count_plan": Object {
                  "description": "计划品项数量",
                  "format": "int32",
                  "type": "integer",
                },
                "item_count_real": Object {
                  "description": "实际品项数量",
                  "format": "int32",
                  "type": "integer",
                },
                "order_id": Object {
                  "description": "补货订单id",
                  "type": "string",
                },
                "status": Object {
                  "description": "状态，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成",
                  "format": "int32",
                  "type": "integer",
                },
                "store_abbr_name": Object {
                  "description": "门店简称",
                  "type": "string",
                },
                "store_code": Object {
                  "description": "门店code",
                  "format": "int32",
                  "type": "integer",
                },
                "store_id": Object {
                  "description": "门店id",
                  "type": "string",
                },
                "store_name": Object {
                  "description": "门店名称",
                  "type": "string",
                },
                "time_create": Object {
                  "description": "创建时间",
                  "format": "int64",
                  "type": "integer",
                },
                "time_plan": Object {
                  "description": "计划到货时间",
                  "format": "int64",
                  "type": "integer",
                },
                "time_real": Object {
                  "description": "实际到货时间",
                  "format": "int64",
                  "type": "integer",
                },
                "time_update": Object {
                  "description": "更新时间",
                  "format": "int64",
                  "type": "integer",
                },
                "user_id_update": Object {
                  "description": "更新人id",
                  "type": "string",
                },
                "user_name_update": Object {
                  "description": "更新人名称",
                  "type": "string",
                },
              },
              "title": "补货需求单，列表用",
              "type": "object",
            },
            "type": "array",
          },
          "type": "array",
        },
        "errcode": Object {
          "format": "int32",
          "type": "integer",
        },
        "errmsg": Object {
          "type": "string",
        },
        "pager": Object {
          "properties": Object {
            "first_item_on_page": Object {
              "format": "int32",
              "type": "integer",
            },
            "has_next_page": Object {
              "type": "boolean",
            },
            "has_previous_page": Object {
              "type": "boolean",
            },
            "is_first_page": Object {
              "type": "boolean",
            },
            "is_last_page": Object {
              "type": "boolean",
            },
            "last_item_on_page": Object {
              "format": "int32",
              "type": "integer",
            },
            "page_count": Object {
              "format": "int32",
              "type": "integer",
            },
            "page_number": Object {
              "format": "int32",
              "type": "integer",
            },
            "page_size": Object {
              "format": "int32",
              "type": "integer",
            },
            "total_item_count": Object {
              "format": "int64",
              "type": "integer",
            },
          },
          "title": "Pager",
          "type": "object",
        },
      },
      "title": "PupuPagerDataResponse«List«补货需求单，列表用»»",
      "type": "object",
    },
    "补货需求单，列表用": Object {
      "properties": Object {
        "apply_no": Object {
          "description": "需求单号",
          "type": "string",
        },
        "apply_type": Object {
          "description": "补货类型,10=旧版采购补货,20=总仓自动补货,30=DC仓采购补货(手动)",
          "format": "int32",
          "type": "integer",
        },
        "date_part_plan": Object {
          "description": "计划到货时间段，10=白天，20=晚上",
          "format": "int32",
          "type": "integer",
        },
        "date_plan": Object {
          "description": "计划到货日期",
          "format": "int64",
          "type": "integer",
        },
        "dc_abbr_name": Object {
          "description": "总仓简称",
          "type": "string",
        },
        "dc_id": Object {
          "description": "总仓id",
          "type": "string",
        },
        "dc_name": Object {
          "description": "总仓名称",
          "type": "string",
        },
        "id": Object {
          "description": "id",
          "type": "string",
        },
        "is_delete": Object {
          "description": "是否删除",
          "example": false,
          "type": "boolean",
        },
        "item_count_plan": Object {
          "description": "计划品项数量",
          "format": "int32",
          "type": "integer",
        },
        "item_count_real": Object {
          "description": "实际品项数量",
          "format": "int32",
          "type": "integer",
        },
        "order_id": Object {
          "description": "补货订单id",
          "type": "string",
        },
        "status": Object {
          "description": "状态，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成",
          "format": "int32",
          "type": "integer",
        },
        "store_abbr_name": Object {
          "description": "门店简称",
          "type": "string",
        },
        "store_code": Object {
          "description": "门店code",
          "format": "int32",
          "type": "integer",
        },
        "store_id": Object {
          "description": "门店id",
          "type": "string",
        },
        "store_name": Object {
          "description": "门店名称",
          "type": "string",
        },
        "time_create": Object {
          "description": "创建时间",
          "format": "int64",
          "type": "integer",
        },
        "time_plan": Object {
          "description": "计划到货时间",
          "format": "int64",
          "type": "integer",
        },
        "time_real": Object {
          "description": "实际到货时间",
          "format": "int64",
          "type": "integer",
        },
        "time_update": Object {
          "description": "更新时间",
          "format": "int64",
          "type": "integer",
        },
        "user_id_update": Object {
          "description": "更新人id",
          "type": "string",
        },
        "user_name_update": Object {
          "description": "更新人名称",
          "type": "string",
        },
      },
      "title": "补货需求单，列表用",
      "type": "object",
    },
  },
  "info": Object {
    "description": "朴朴java服务端接口文档，遵循RESTful规范",
    "title": "admin-scm",
    "version": "0.0.1",
  },
  "paths": Object {
    "/admin/depot/apply": Object {
      "get": Object {
        "operationId": "pageApplyUsingGET",
        "parameters": Array [
          Object {
            "description": "用户签名，JWT Bearer",
            "in": "header",
            "name": "Authorization",
            "required": false,
            "type": "string",
          },
          Object {
            "description": "需求单号",
            "in": "query",
            "name": "applyNo",
            "required": false,
            "type": "string",
          },
          Object {
            "description": "门店id",
            "in": "query",
            "name": "storeId",
            "required": false,
            "type": "string",
          },
          Object {
            "description": "状态,-1=全部，1=草稿，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成 ，120=未完成  200=删除",
            "format": "int32",
            "in": "query",
            "name": "status",
            "required": false,
            "type": "integer",
          },
          Object {
            "description": "创建时间开始",
            "format": "int64",
            "in": "query",
            "name": "timeCreateFrom",
            "required": false,
            "type": "integer",
          },
          Object {
            "description": "创建时间结束",
            "format": "int64",
            "in": "query",
            "name": "timeCreateTo",
            "required": false,
            "type": "integer",
          },
          Object {
            "default": 0,
            "description": "第几页(0..N)",
            "format": "int32",
            "in": "query",
            "name": "page",
            "required": true,
            "type": "integer",
          },
          Object {
            "default": 20,
            "description": "每页大小",
            "format": "int32",
            "in": "query",
            "name": "size",
            "required": true,
            "type": "integer",
          },
          Object {
            "collectionFormat": "multi",
            "description": "排序，格式: property(,asc|desc). 默认升序. ",
            "in": "query",
            "items": Object {
              "type": "string",
            },
            "name": "sort",
            "required": false,
            "type": "array",
          },
        ],
        "produces": Array [
          "*/*",
        ],
        "responses": Object {
          "200": Object {
            "description": "OK",
            "schema": Object {
              "properties": Object {
                "count": Object {
                  "format": "int64",
                  "type": "integer",
                },
                "data": Object {
                  "items": Object {
                    "items": Object {
                      "properties": Object {
                        "apply_no": Object {
                          "description": "需求单号",
                          "type": "string",
                        },
                        "apply_type": Object {
                          "description": "补货类型,10=旧版采购补货,20=总仓自动补货,30=DC仓采购补货(手动)",
                          "format": "int32",
                          "type": "integer",
                        },
                        "date_part_plan": Object {
                          "description": "计划到货时间段，10=白天，20=晚上",
                          "format": "int32",
                          "type": "integer",
                        },
                        "date_plan": Object {
                          "description": "计划到货日期",
                          "format": "int64",
                          "type": "integer",
                        },
                        "dc_abbr_name": Object {
                          "description": "总仓简称",
                          "type": "string",
                        },
                        "dc_id": Object {
                          "description": "总仓id",
                          "type": "string",
                        },
                        "dc_name": Object {
                          "description": "总仓名称",
                          "type": "string",
                        },
                        "id": Object {
                          "description": "id",
                          "type": "string",
                        },
                        "is_delete": Object {
                          "description": "是否删除",
                          "example": false,
                          "type": "boolean",
                        },
                        "item_count_plan": Object {
                          "description": "计划品项数量",
                          "format": "int32",
                          "type": "integer",
                        },
                        "item_count_real": Object {
                          "description": "实际品项数量",
                          "format": "int32",
                          "type": "integer",
                        },
                        "order_id": Object {
                          "description": "补货订单id",
                          "type": "string",
                        },
                        "status": Object {
                          "description": "状态，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成",
                          "format": "int32",
                          "type": "integer",
                        },
                        "store_abbr_name": Object {
                          "description": "门店简称",
                          "type": "string",
                        },
                        "store_code": Object {
                          "description": "门店code",
                          "format": "int32",
                          "type": "integer",
                        },
                        "store_id": Object {
                          "description": "门店id",
                          "type": "string",
                        },
                        "store_name": Object {
                          "description": "门店名称",
                          "type": "string",
                        },
                        "time_create": Object {
                          "description": "创建时间",
                          "format": "int64",
                          "type": "integer",
                        },
                        "time_plan": Object {
                          "description": "计划到货时间",
                          "format": "int64",
                          "type": "integer",
                        },
                        "time_real": Object {
                          "description": "实际到货时间",
                          "format": "int64",
                          "type": "integer",
                        },
                        "time_update": Object {
                          "description": "更新时间",
                          "format": "int64",
                          "type": "integer",
                        },
                        "user_id_update": Object {
                          "description": "更新人id",
                          "type": "string",
                        },
                        "user_name_update": Object {
                          "description": "更新人名称",
                          "type": "string",
                        },
                      },
                      "title": "补货需求单，列表用",
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "type": "array",
                },
                "errcode": Object {
                  "format": "int32",
                  "type": "integer",
                },
                "errmsg": Object {
                  "type": "string",
                },
                "pager": Object {
                  "properties": Object {
                    "first_item_on_page": Object {
                      "format": "int32",
                      "type": "integer",
                    },
                    "has_next_page": Object {
                      "type": "boolean",
                    },
                    "has_previous_page": Object {
                      "type": "boolean",
                    },
                    "is_first_page": Object {
                      "type": "boolean",
                    },
                    "is_last_page": Object {
                      "type": "boolean",
                    },
                    "last_item_on_page": Object {
                      "format": "int32",
                      "type": "integer",
                    },
                    "page_count": Object {
                      "format": "int32",
                      "type": "integer",
                    },
                    "page_number": Object {
                      "format": "int32",
                      "type": "integer",
                    },
                    "page_size": Object {
                      "format": "int32",
                      "type": "integer",
                    },
                    "total_item_count": Object {
                      "format": "int64",
                      "type": "integer",
                    },
                  },
                  "title": "Pager",
                  "type": "object",
                },
              },
              "title": "PupuPagerDataResponse«List«补货需求单，列表用»»",
              "type": "object",
            },
          },
          "401": Object {
            "description": "Unauthorized",
          },
          "403": Object {
            "description": "Forbidden",
          },
          "404": Object {
            "description": "Not Found",
          },
        },
        "summary": "分页查找补货需求单: 3610401",
        "tags": Array [
          "depot-apply-controller",
        ],
      },
    },
  },
  "swagger": "2.0",
}
`;
